### ISSUE TITLE
Subscriptions - Websocket keeps disconnecting constantly

### ISSUE THREAD

**Issue by Rahuvich** [zino-app/graphql-flutter#729]
**Issue**
Websocket keeps disconnecting constantly.

```dart
flutter: Connected to websocket.
flutter: PARAMS null
flutter: Disconnected from websocket.
flutter: Scheduling to connect in 1 seconds...
flutter: Connecting to websocket: ws://myweb.dev:3030/graphql...
flutter: Connected to websocket.
flutter: Disconnected from websocket.
flutter: Scheduling to connect in 1 seconds...
flutter: PARAMS null
flutter: Connecting to websocket: ws://myweb.dev:3030/graphql...
flutter: Connected to websocket.
flutter: PARAMS null
flutter: Disconnected from websocket.
flutter: Scheduling to connect in 1 seconds...
flutter: Connecting to websocket: ws://myweb.dev:3030/graphql...
flutter: Connected to websocket.
flutter: PARAMS null
flutter: Disconnected from websocket.
flutter: Scheduling to connect in 1 seconds...
flutter: Connecting to websocket: ws://myweb.dev:3030/graphql...
flutter: Connected to websocket.
```

WebSocket Repository
```dart

class WebSocketRepository {
  static messageSent(roomId) => gql("""
   subscription messageSent {
   messageSent(roomId: "$roomId") {
    id: _id
    text
    ad {
      ...adFields
    }
    sender {
      ...userFields
    }
    createdAt
    updatedAt
   }
 }
 """ +
      AdsProvider.adFragment);

  static AuthLink authLink(token) =>
      AuthLink(getToken: () async => 'Bearer $token');

  static WebSocketLink get websocketLink => WebSocketLink(
        'ws://' + ApiBaseHelper.domain,
        config: SocketClientConfig(
          autoReconnect: true,
          inactivityTimeout: Duration(hours: 1),
          delayBetweenReconnectionAttempts: Duration(seconds: 1),
        ),
      );

  static ValueNotifier<GraphQLClient> initailizeClient(String token) {
    print('token $token');
    ValueNotifier<GraphQLClient> client = ValueNotifier(
      GraphQLClient(
        cache: GraphQLCache(store: HiveStore()),
        link: authLink(token).concat(websocketLink),
      ),
    );
    return client;
  }
}
```

Widget
```dart
class ChatPage extends StatelessWidget {
  final Room room;
  ChatPage({@required this.room});
  @override
  Widget build(BuildContext context) {
    return GraphQLProvider(
        client: WebSocketRepository.initailizeClient,
        child: Subscription(
          options: SubscriptionOptions(
              document: WebSocketRepository.messageSent(room.id)),
          builder: (result) {
            print('PARAMS $result');

            return _InnerChatPage(
              room: room,
            );
          },
        ));
  }
}
```

Parameter result is always null.

**Context**
Version 4.0.0-alpha.6
Running in a iOS simulator
```
$ flutter doctor -v
[✓] Flutter (Channel stable, 1.20.1, on Mac OS X 10.15.6 19G2021,
    locale es-ES)
    • Flutter version 1.20.1 at /Users/raulmateobeneyto/flutter
    • Framework revision 2ae34518b8 (6 weeks ago), 2020-08-05
      19:53:19 -0700
    • Engine revision c8e3b94853
    • Dart version 2.9.0

 
[✓] Android toolchain - develop for Android devices (Android SDK 
    version 29.0.2)
    • Android SDK at /Users/raulmateobeneyto/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at:
      /Library/Java/JavaVirtualMachines/jdk1.8.0_192.jdk/Contents/
      Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build
      1.8.0_192-b12)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4, Build version 11E146
    • CocoaPods version 1.8.4

[!] Android Studio (not installed)
    • Android Studio not found; download from
      https://developer.android.com/studio/index.html
      (or visit
      https://flutter.dev/docs/get-started/install/macos#android-s
      etup for detailed instructions).

[✓] VS Code (version 1.49.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (2 available)
    • iPhone 11 Pro (mobile)     •
      D2CC28C5-236E-449A-A0F3-5B96652F3D58 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-13-4 (simulator)
    • iPhone 11 Pro Max (mobile) •
      7E5602B3-666D-443E-BCB8-32A41CE1FF10 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-13-4 (simulator)

! Doctor found issues in 1 category.
```

**Comment by micimize** [comment]
how quickly does this happen? Nothing looks wrong with your code – my guess is that the server is actually terminating the connection gracefully, as no errors are logged.

**Comment by HofmannZ** [comment]
Does your server have keep-alive setup? I know Apollo server does not have this out of the box.

```ts
new ApolloServer({
  ...
  subscriptions: {
    keepAlive: 1000, // 10 seconds
  },
});
```

**Comment by Rahuvich** [comment]
I checked deeper and the problem is my backend doesn't receive a single header, therefore it doesn't get my user token. What's the way of passing auth token on a websocket link? I tried to concat AuthLink with WebsocketLink with no luck.

**Comment by HofmannZ** [comment]
@Rahuvich you can use something like this:

```dart
  final websocketLink = WebSocketLink(
    'ws://localhost:3001/',
    config: SocketClientConfig(
      initialPayload: () async {
        final token = await getToken();

        return {
          'Authorization': 'Bearer $token',
        };
      },
    ),
  );
```

**Comment by kateile** [comment]
Hi @HofmannZ . My server doesnt receive `cookie` parameter with solution below

```dart
  final _wsLink = WebSocketLink(
    'ws://$url',
    config: SocketClientConfig(
      //inactivityTimeout: Duration(minutes: 30),
      delayBetweenReconnectionAttempts: Duration(hours: 1),
      autoReconnect: !url.contains('9002'),
      initialPayload: () async {
        final List<String> cookies = cookieJar
            .loadForRequest(Uri.parse(graphqlEndpoint))
            .where((c) => c.value.length >= 20)
            .map((e) => '${e.name}=${e.value}')
            .toList();

        final String cookie = cookies.join('; ');

        return {'cookie': cookie};
      },
    ),
  );
```

**Comment by DervisAvdic** [comment]
@Arjun544 Thank you very much. 

`subProtocol: GraphQLProtocol.graphqlTransportWs, // This line is important, by default protocol is GraphQLProtocol.graphqlWs
`
this was missing in my code and because of it the subscription couldn't start. My subscription did start from Apollo playground but I got this error 

Disconnected (code: 4406, reason: "Subprotocol not acceptable") --> Seems like subprotocols was not correct and that did the trick. I tried with this however, 

`'headers': {
            'Sec-WebSocket-Protocol': 'graphql-ws',
          },`  
But it didn't work.
